iothub/device/src/ConnectionInfo.cs

```diff
@@ -36,13 +36,51 @@ internal ConnectionInfo(ConnectionStatus status, ConnectionStatusChangeReason ch
         /// The reason for the current connection status change.
         /// </summary>
         /// <remark>
-        /// Defaults to <see cref="ConnectionStatusChangeReason.ClientClose"/>.
+        /// Defaults to <see cref="ConnectionStatusChangeReason.ClientClosed"/>.
         /// </remark>
         public ConnectionStatusChangeReason ChangeReason { get; }
 
         /// <summary>
         /// Timestamp in UTC when the last connection status was changed.
         /// </summary>
-        public DateTimeOffset StatusLastChangedOnUtc { get; internal set; }
+        public DateTimeOffset StatusLastChangedOnUtc { get; }
+
+        /// <summary>
+        /// Recommended actions for users to take upon different ConnectionStatus and ConnectionStatusChangeReason.
+        /// </summary>
+        /// <remark>
+        /// Defaults to <see cref="RecommendedAction.OpenConnection"/>.
+        /// </remark>>
+        public RecommendedAction RecommendedAction { get; }
+
+        private static RecommendedAction GetRecommendedAction(ConnectionStatus status, ConnectionStatusChangeReason changeReason)
+        {
+            switch (status)
+            {
+                case ConnectionStatus.Connected:
+                    return RecommendedAction.PerformNormally;
+
+                case ConnectionStatus.DisconnectedRetrying:
+                    return RecommendedAction.WaitForRetryPolicy;
+
+                case ConnectionStatus.Closed:
+                    return RecommendedAction.Quit;
+
+                case ConnectionStatus.Disconnected:
+                    switch (changeReason)
+                    {
+                        case ConnectionStatusChangeReason.RetryExpired:
+                        case ConnectionStatusChangeReason.CommunicationError:
+                            return RecommendedAction.OpenConnection;
+
+                        case ConnectionStatusChangeReason.BadCredential:
+                        case ConnectionStatusChangeReason.DeviceDisabled:
+                            return RecommendedAction.Quit;
+                    }
+                    break;
+            }
+
+            return RecommendedAction.OpenConnection;
```

Comment:
Q - other than disconnected-retry_expired, communication_error, what other status and reason combination gives us the Open recommended action?

Reply:
these 2 should be all. `OpenConnection` is also the default recommended action when a client is first initialized though

Reply:
you could add a comment calling that out in this sample